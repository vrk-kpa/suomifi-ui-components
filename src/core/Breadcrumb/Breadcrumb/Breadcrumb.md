The `<Breadcrumb>` component is used to let the user know their current location in a web service. It is rendered as an HTML `nav` element.

### Accessibility Notes

- Height does not adjust when text wraps, e.g. when text is resized 200%.

Examples:

- [Basic use](./#/Components/Breadcrumb?id=basic-use)

<div style="margin-bottom: 5px">
 [Props & methods (Breadcrumb)](./#/Components/Breadcrumb?id=props--methods)
</div>
<div style="margin-bottom: 40px">
  [Props & methods (BreadcrumbLink)](./#/Components/Breadcrumb?id=breadcrumblink)
</div>

### Basic use

Use an `aria-label` to describe the purpose of the Breadcrumb element and distinguish it from other navigation landmarks on the same page.

Mark the current page with the `current` prop.

```jsx
import { Breadcrumb, BreadcrumbLink } from 'suomifi-ui-components';

<Breadcrumb aria-label="Your current location">
  <BreadcrumbLink href="/">Home</BreadcrumbLink>
  <BreadcrumbLink href="/citizens">Citizens</BreadcrumbLink>
  <BreadcrumbLink current>Social security</BreadcrumbLink>
</Breadcrumb>;
```

### Props & methods

Breadcrumb component supports [margin props](./#/Spacing/Margin%20props) for spacing.
