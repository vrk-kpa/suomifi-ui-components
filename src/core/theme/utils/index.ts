export {
  withSuomifiDefaultProps,
  themeOrTokens,
  TokensOrThemeProps,
} from './defaults';
export { focus } from './focus';
export { boxshadowOutline } from './outline';
export { margin, padding, spacingModifiers } from './spacing';
export { typographyUtils, fontSize } from './typography';
