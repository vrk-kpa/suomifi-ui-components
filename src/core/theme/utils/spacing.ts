import {
  SuomifiTheme,
  SpacingProp,
  defaultSuomifiTheme,
} from '../SuomifiTheme/SuomifiTheme';

export type SpacingWithoutInsetProp =
  | 'xxs'
  | 'xs'
  | 's'
  | 'm'
  | 'l'
  | 'xl'
  | 'xxl'
  | 'xxxl'
  | 'xxxxl'
  | '0';

export type GlobalMarginProps = { globalMargins: GlobalMargins };

const spaceVal = (theme: SuomifiTheme) => (val?: SpacingProp) => {
  if (val === '0') return '0';
  return !!val ? theme.spacing[val] : '';
};

export interface PaddingProps {
  /** Padding from theme */
  padding?: SpacingWithoutInsetProp;
  /** Padding-top from theme */
  pt?: SpacingWithoutInsetProp;
  /** Padding-right from theme */
  pr?: SpacingWithoutInsetProp;
  /** Padding-bottom from theme */
  pb?: SpacingWithoutInsetProp;
  /** Padding-left from theme */
  pl?: SpacingWithoutInsetProp;
  /** Padding on the x-axis (left & right) from theme */
  px?: SpacingWithoutInsetProp;
  /** Padding on the y-axis (top & bottom) from theme */
  py?: SpacingWithoutInsetProp;
}
export interface MarginProps {
  /** Margin from theme */
  margin?: SpacingWithoutInsetProp;
  /** Margin-top from theme */
  mt?: SpacingWithoutInsetProp;
  /** Margin-right from theme */
  mr?: SpacingWithoutInsetProp;
  /** Margin-bottom from theme */
  mb?: SpacingWithoutInsetProp;
  /** Margin-left from theme */
  ml?: SpacingWithoutInsetProp;
  /** Margin on the x-axis (left & right) from theme */
  mx?: SpacingWithoutInsetProp;
  /** Margin on the y-axis (top & bottom) from theme */
  my?: SpacingWithoutInsetProp;
}

export type GlobalMargins = {
  all?: MarginProps;
  alert?: MarginProps;
  actionMenu?: MarginProps;
  block?: MarginProps;
  button?: MarginProps;
  breadcrumb?: MarginProps;
  checkbox?: MarginProps;
  checkboxGroup?: MarginProps;
  chip?: MarginProps;
  dateInput?: MarginProps;
  dropdown?: MarginProps;
  expander?: MarginProps;
  expanderGroup?: MarginProps;
  externalLink?: MarginProps;
  fileInput?: MarginProps;
  heading?: MarginProps;
  hintText?: MarginProps;
  inlineAlert?: MarginProps;
  label?: MarginProps;
  languageMenu?: MarginProps;
  link?: MarginProps;
  linkList?: MarginProps;
  loadingSpinner?: MarginProps;
  multiSelect?: MarginProps;
  notification?: MarginProps;
  pagination?: MarginProps;
  paragraph?: MarginProps;
  radioButton?: MarginProps;
  radioButtonGroup?: MarginProps;
  routerLink?: MarginProps;
  searchInput?: MarginProps;
  serviceNavigation?: MarginProps;
  sideNavigation?: MarginProps;
  singleSelect?: MarginProps;
  staticChip?: MarginProps;
  statusText?: MarginProps;
  text?: MarginProps;
  textarea?: MarginProps;
  textInput?: MarginProps;
  timeInput?: MarginProps;
  toast?: MarginProps;
  toggleInput?: MarginProps;
  toggleButton?: MarginProps;
  tooltip?: MarginProps;
  wizardNavigation?: MarginProps;
};

export interface SpacingProps extends PaddingProps, MarginProps {}

export const spacingStyles = (props: SpacingProps | undefined) => {
  if (!props) return;
  const array = Object.entries(props).map(([key, value]) =>
    getSpacingStyle(defaultSuomifiTheme, key as keyof SpacingProps, value),
  );
  return Object.assign({}, ...array);
};

export const buildSpacingCSS = (
  spacing?: SpacingProps,
  important?: boolean,
): string => {
  if (!spacing) return '';

  const cssStyles = Object.entries(spacing)
    .map(([key, value]) =>
      getCSSSpacing(
        defaultSuomifiTheme,
        key as keyof SpacingProps,
        value,
        important,
      ),
    )
    .join('');

  return cssStyles;
};

const inlineStyle = {
  margin: 'margin',
  mt: 'marginTop',
  mr: 'marginRight',
  mb: 'marginBottom',
  ml: 'marginLeft',
  padding: 'padding',
  pt: 'paddingTop',
  pr: 'paddingRight',
  pb: 'paddingBottom',
  pl: 'paddingLeft',
};

const cssSelector = {
  margin: 'margin',
  mt: 'margin-top',
  mr: 'margin-right',
  mb: 'margin-bottom',
  ml: 'margin-left',
  padding: 'padding',
  pt: 'padding-top',
  pr: 'padding-right',
  pb: 'padding-bottom',
  pl: 'padding-left',
};

const getSpacingStyle = (
  theme: SuomifiTheme,
  key: keyof SpacingProps,
  value: SpacingProp,
) => {
  const amount = spaceVal(theme)(value);
  switch (key) {
    case 'mx':
      return { marginRight: `${amount}`, marginLeft: `${amount}` };
    case 'my':
      return { marginTop: `${amount}`, marginBottom: `${amount}` };
    case 'px':
      return { paddingRight: `${amount}`, paddingLeft: `${amount}` };
    case 'py':
      return { paddingTop: `${amount}`, paddingBottom: `${amount}` };
    case 'margin':
    case 'mt':
    case 'mr':
    case 'mb':
    case 'ml':
    case 'padding':
    case 'pt':
    case 'pr':
    case 'pb':
    case 'pl':
      return { [inlineStyle[key]]: `${amount}` };
    default:
      return '';
  }
};

const getCSSSpacing = (
  theme: SuomifiTheme,
  key: keyof SpacingProps,
  value: SpacingProp,
  important: boolean = false,
) => {
  const amount = spaceVal(theme)(value);
  const importantValue = important ? '!important' : '';
  switch (key) {
    case 'mx':
      return `margin-right: ${amount} ${importantValue}; margin-left: ${amount} ${importantValue};`;
    case 'my':
      return `margin-top: ${amount} ${importantValue}; margin-bottom: ${amount} ${importantValue};`;
    case 'px':
      return `padding-right: ${amount} ${importantValue}; padding-left: ${amount} ${importantValue};`;
    case 'py':
      return `padding-top: ${amount} ${importantValue}; padding-bottom: ${amount} ${importantValue};`;
    default:
      return `${cssSelector[key]}: ${amount} ${importantValue};`;
  }
};

export const separateMarginProps = <T extends MarginProps>(
  props: T,
): [MarginProps, Omit<T, keyof MarginProps>] => {
  const { margin, my, mx, mt, mr, mb, ml, ...otherProps } = props;
  return [
    {
      ...(margin !== undefined && { margin }),
      ...(my !== undefined && { my }),
      ...(mx !== undefined && { mx }),
      ...(mt !== undefined && { mt }),
      ...(mr !== undefined && { mr }),
      ...(mb !== undefined && { mb }),
      ...(ml !== undefined && { ml }),
    },
    otherProps,
  ];
};

export const separateMarginAndPaddingProps = <T extends SpacingProps>(
  props: T,
): [SpacingProps, Omit<T, keyof SpacingProps>] => {
  const {
    margin,
    padding,
    my,
    mx,
    py,
    px,
    mt,
    mr,
    mb,
    ml,
    pt,
    pr,
    pb,
    pl,
    ...otherProps
  } = props;
  return [
    {
      ...(margin !== undefined && { margin }),
      ...(padding !== undefined && { padding }),
      ...(my !== undefined && { my }),
      ...(mx !== undefined && { mx }),
      ...(py !== undefined && { py }),
      ...(px !== undefined && { px }),
      ...(mt !== undefined && { mt }),
      ...(mr !== undefined && { mr }),
      ...(mb !== undefined && { mb }),
      ...(ml !== undefined && { ml }),
      ...(pt !== undefined && { pt }),
      ...(pr !== undefined && { pr }),
      ...(pb !== undefined && { pb }),
      ...(pl !== undefined && { pl }),
    },
    otherProps,
  ];
};
