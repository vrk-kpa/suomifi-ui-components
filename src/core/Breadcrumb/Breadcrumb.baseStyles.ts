import { css } from 'styled-components';
import { withSuomifiTheme, TokensAndTheme } from '../theme';
import { nav, list, listItem, font } from '../theme/reset';
import { fontSize } from '../theme/utils';

export const baseStyles = withSuomifiTheme(
  ({ theme }: TokensAndTheme) => css`
  ${nav({ theme })}
  ${font({ theme })('bodyText')}
  background-color: ${theme.colors.whiteBase};

  & .fi-breadcrumb {
    &_list {
      ${list({ theme })}
      ${font({ theme })('bodyText')}
      margin: 0;
      padding: 0;
    }
    &_item {
      ${listItem({ theme })}
      ${font({ theme })('bodyText')}
      float: left;
    }
    &_item,
    &_link,
    &_icon {
      font-size: ${fontSize({ theme })('bodyText')};
    }
    &_icon {
      transform: translateY(.2em);
    }
  }
`,
);
