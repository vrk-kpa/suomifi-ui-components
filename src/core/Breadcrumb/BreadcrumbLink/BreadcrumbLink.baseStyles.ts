import { css } from 'styled-components';
import { SuomifiTheme } from '../../theme';
import { font } from '../../theme/reset';
import { cssValueToString } from '../../../utils/css';

export const baseStyles = (theme: SuomifiTheme) => css`
  font-size: ${cssValueToString(
    theme.values.typography.bodyTextSmall.fontSize,
  )};
  &.fi-breadcrumb-link {
    display: inline-block;

    & .fi-breadcrumb-link_link {
      ${font(theme)('bodyTextSmall')}
      margin: 0;

      &:hover,
      &:active {
        text-decoration: underline;
      }

      &--initial-underline {
        text-decoration: underline;

        &:focus,
        &:focus-within {
          text-decoration: underline;
        }

        &:hover,
        &:active {
          text-decoration: none;
        }
      }

      &--current {
        color: ${theme.colors.depthDark1};
        text-decoration: none;

        &:hover,
        &:active {
          text-decoration: none;
        }
      }
    }

    & .fi-breadcrumb-link_icon {
      transform: translateY(0.2em);
      margin: 0 ${theme.spacing.insetXs};
      & .fi-icon-base-fill {
        fill: ${theme.colors.depthDark1};
      }
    }
  }
`;
