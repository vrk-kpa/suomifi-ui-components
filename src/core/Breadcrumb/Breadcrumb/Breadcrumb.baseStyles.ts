import { css } from 'styled-components';
import { SuomifiTheme } from '../../theme';
import { nav, list, listItem, font } from '../../theme/reset';
import { MarginProps, buildSpacingCSS } from '../../theme/utils/spacing';

export const baseStyles = (
  theme: SuomifiTheme,
  globalMargins?: MarginProps,
  propMargins?: MarginProps,
) => css`
  ${nav(theme)}
  ${font(theme)('bodyTextSmall')}
  ${buildSpacingCSS(globalMargins)}
  ${buildSpacingCSS(propMargins, true)};
  height: 1.5em;

  & .fi-breadcrumb {
    &_list {
      ${list(theme)}
      margin: 0;
      padding: 0;
    }
    &_item {
      ${listItem(theme)}
      float: left;
    }
  }
`;
