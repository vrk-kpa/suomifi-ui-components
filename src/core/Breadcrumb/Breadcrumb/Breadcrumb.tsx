import React, { Component, ReactNode } from 'react';
import { default as styled } from 'styled-components';
import classnames from 'classnames';
import { getConditionalAriaProp } from '../../../utils/aria';
import { HtmlLi, HtmlNav, HtmlNavProps, HtmlOl } from '../../../reset';
import { baseStyles } from './Breadcrumb.baseStyles';
import {
  SuomifiThemeProp,
  SuomifiThemeConsumer,
  SpacingConsumer,
} from '../../theme';
import {
  separateMarginProps,
  MarginProps,
  GlobalMarginProps,
} from '../../theme/utils/spacing';
import { filterDuplicateKeys } from '../../../utils/common/common';

const baseClassName = 'fi-breadcrumb';
const listClassName = `${baseClassName}_list`;
const itemClassName = `${baseClassName}_item`;

export interface BreadcrumbProps extends HtmlNavProps, MarginProps {
  /** Labels the breadcrumb for screen reader users */
  'aria-label': string;
  /** CSS class for custom styles */
  className?: string;
  /**
   * Use `<BreadcrumbLink>` elements as children
   */
  children?: ReactNode;
}

const breadcrumbItems = (children: ReactNode) =>
  React.Children.map(children, (child) => (
    <HtmlLi className={itemClassName}>{child}</HtmlLi>
  ));

class BaseBreadcrumb extends Component<BreadcrumbProps> {
  render() {
    const { className, children, ...rest } = this.props;
    const [_marginProps, passProps] = separateMarginProps(rest);

    return (
      <HtmlNav
        {...passProps}
        className={classnames(baseClassName, className)}
        style={{ ...passProps?.style }}
      >
        <HtmlOl className={listClassName}>{breadcrumbItems(children)}</HtmlOl>
      </HtmlNav>
    );
  }
}

const StyledBreadcrumb = styled(
  ({
    globalMargins,
    theme,
    ...passProps
  }: BreadcrumbProps & SuomifiThemeProp & GlobalMarginProps) => (
    <BaseBreadcrumb {...passProps} />
  ),
)`
  ${({ theme, globalMargins, ...rest }) => {
    const [marginProps, _passProps] = separateMarginProps(rest);
    const cleanedGlobalMargins = filterDuplicateKeys(
      globalMargins.breadcrumb,
      marginProps,
    );
    return baseStyles(theme, cleanedGlobalMargins, marginProps);
  }}
`;

const Breadcrumb = (props: BreadcrumbProps) => {
  const { 'aria-label': ariaLabel, ...passProps } = props;
  return (
    <SpacingConsumer>
      {({ margins }) => (
        <SuomifiThemeConsumer>
          {({ suomifiTheme }) => (
            <StyledBreadcrumb
              theme={suomifiTheme}
              globalMargins={margins}
              {...passProps}
              {...getConditionalAriaProp('aria-label', [ariaLabel])}
            />
          )}
        </SuomifiThemeConsumer>
      )}
    </SpacingConsumer>
  );
};

Breadcrumb.displayName = 'Breadcrumb';
export { Breadcrumb };
