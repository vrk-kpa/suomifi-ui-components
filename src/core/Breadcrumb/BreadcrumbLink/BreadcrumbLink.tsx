import React from 'react';
import { styled } from 'styled-components';
import { Link, LinkProps } from '../../Link';
import { HtmlSpan } from '../../../reset';
import classnames from 'classnames';
import { baseStyles } from './BreadcrumbLink.baseStyles';
import { SuomifiThemeConsumer, SuomifiThemeProp } from '../../theme';
import { IconLinkBreadcrumb } from 'suomifi-icons';

const baseClassName = 'fi-breadcrumb-link';
const breadcrumbClassNames = {
  link: `${baseClassName}_link`,
  current: `${baseClassName}_link--current`,
  linkUnderline: `${baseClassName}_link--initial-underline`,
  icon: `${baseClassName}_icon`,
};

export interface BreadcrumbLinkProps
  extends Omit<
    LinkProps,
    'href' | 'margin' | 'mx' | 'my' | 'mt' | 'mr' | 'mb' | 'ml'
  > {
  /** Indicates the link is the current page */
  current?: boolean;
  /** Url for the link */
  href?: string;
}

const BaseBreadcrumbLink = (props: BreadcrumbLinkProps & SuomifiThemeProp) => {
  const {
    current,
    children,
    className,
    theme,
    underline = 'hover',
    href = '',
    ...passProps
  } = props;
  return (
    <HtmlSpan className={classnames(className, baseClassName)}>
      {!current ? (
        <>
          <Link
            {...passProps}
            className={classnames(breadcrumbClassNames.link, {
              [breadcrumbClassNames.linkUnderline]: underline === 'initial',
            })}
            href={href}
          >
            {children}
          </Link>
          <IconLinkBreadcrumb className={breadcrumbClassNames.icon} />
        </>
      ) : (
        <HtmlSpan
          className={classnames(
            breadcrumbClassNames.link,
            breadcrumbClassNames.current,
          )}
          aria-current="location"
        >
          {children}
        </HtmlSpan>
      )}
    </HtmlSpan>
  );
};

const StyledBreadcrumbLink = styled(BaseBreadcrumbLink)`
  ${({ theme }) => baseStyles(theme)}
`;

export const BreadcrumbLink = (props: BreadcrumbLinkProps) => (
  <SuomifiThemeConsumer>
    {({ suomifiTheme }) => (
      <StyledBreadcrumbLink theme={suomifiTheme} {...props} />
    )}
  </SuomifiThemeConsumer>
);
