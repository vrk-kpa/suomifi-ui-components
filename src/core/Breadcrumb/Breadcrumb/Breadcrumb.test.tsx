import React from 'react';
import { render } from '@testing-library/react';
import { axeTest } from '../../../utils/test';
import { Breadcrumb } from './Breadcrumb';
import { BreadcrumbLink } from '../BreadcrumbLink/BreadcrumbLink';

const TestBreadcrumb = (
  <Breadcrumb data-testid="Breadcrumb" aria-label="breadcrumb">
    <BreadcrumbLink href="/">Frontpage</BreadcrumbLink>
    <BreadcrumbLink href="/sub">Sub page</BreadcrumbLink>
    <BreadcrumbLink current={true}>Sub sub page</BreadcrumbLink>
  </Breadcrumb>
);

describe('margin prop', () => {
  it('should have margin style from margin prop', () => {
    const { container } = render(
      <Breadcrumb aria-label="You are here" margin="xs" />,
    );
    const navigation = container.querySelector('nav');
    expect(navigation).toHaveStyle('margin: 10px');
  });

  it('should have margin prop style overwritten from style', () => {
    const { container } = render(
      <Breadcrumb
        aria-label="You are here"
        margin="xs"
        style={{ margin: 2 }}
      />,
    );
    const navigation = container.querySelector('nav');
    expect(navigation).toHaveAttribute('style', 'margin: 2px;');
  });
});

test('calling render with the same component on the same container does not remount', () => {
  const BreadcrumbRendered = render(TestBreadcrumb);
  const { container } = BreadcrumbRendered;
  expect(container.firstChild).toMatchSnapshot();
});

test('should not have basic accessibility issues', axeTest(TestBreadcrumb));
