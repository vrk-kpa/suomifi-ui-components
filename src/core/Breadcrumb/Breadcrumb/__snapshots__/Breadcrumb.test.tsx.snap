// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calling render with the same component on the same container does not remount 1`] = `
.c6 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  background-color: transparent;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c6::before,
.c6::after {
  box-sizing: border-box;
}

.c6:hover,
.c6:active,
.c6:focus,
.c6:focus-within {
  color: inherit;
  -webkit-text-decoration: underline;
  text-decoration: underline;
  cursor: pointer;
}

.c3 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: list-item;
}

.c3::before,
.c3::after {
  box-sizing: border-box;
}

.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
}

.c0::before,
.c0::after {
  box-sizing: border-box;
}

.c2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  list-style-type: decimal;
  margin-top: 1em;
  margin-bottom: 1em;
  margin-left: 0;
  margin-right: 0;
  padding-left: 40px;
}

.c2::before,
.c2::after {
  box-sizing: border-box;
}

.c4 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c4::before,
.c4::after {
  box-sizing: border-box;
}

.c1 {
  color: hsl(0,0%,13%);
  display: block;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
  height: 1.5em;
}

.c1 .fi-breadcrumb_list {
  color: hsl(0,0%,13%);
  list-style: none;
  margin: 0;
  padding: 0;
}

.c1 .fi-breadcrumb_item {
  color: hsl(0,0%,13%);
  list-style: none;
  float: left;
}

.c8 {
  vertical-align: baseline;
}

.c8.fi-icon {
  display: inline-block;
}

.c8 .fi-icon-base-fill {
  fill: currentColor;
}

.c8 .fi-icon-base-stroke {
  stroke: currentColor;
}

.c8.fi-icon--cursor-pointer {
  cursor: pointer;
}

.c8.fi-icon--cursor-pointer * {
  cursor: inherit;
}

.c7 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 18px;
  line-height: 1.5;
  font-weight: 400;
}

.c7.fi-link {
  color: hsl(212,63%,45%);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c7.fi-link:hover,
.c7.fi-link:active,
.c7.fi-link:focus,
.c7.fi-link:focus-within {
  color: hsl(212,63%,45%);
}

.c7.fi-link:focus,
.c7.fi-link:focus-within {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c7.fi-link:focus {
  border-radius: 2px;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  outline: 0;
  outline: 3px solid transparent;
}

.c7.fi-link:hover,
.c7.fi-link:active {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c7.fi-link:visited {
  color: hsl(284,36%,45%);
}

.c7.fi-link--initial-underline {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c7.fi-link--initial-underline:focus,
.c7.fi-link--initial-underline:focus-within {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c7.fi-link--initial-underline:hover,
.c7.fi-link--initial-underline:active {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c7.fi-link.fi-link--accent .fi-link--accent_icon {
  font-size: 16px;
  margin-left: -3px;
  margin-right: 2px;
  -webkit-transform: translateY(0.15em);
  -ms-transform: translateY(0.15em);
  transform: translateY(0.15em);
}

.c7.fi-link.fi-link--small {
  font-size: 16px;
}

.c5 {
  font-size: 16px;
}

.c5.fi-breadcrumb-link {
  display: inline-block;
}

.c5.fi-breadcrumb-link .fi-breadcrumb-link_link {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
  margin: 0;
}

.c5.fi-breadcrumb-link .fi-breadcrumb-link_link:hover,
.c5.fi-breadcrumb-link .fi-breadcrumb-link_link:active {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c5.fi-breadcrumb-link .fi-breadcrumb-link_link--initial-underline {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c5.fi-breadcrumb-link .fi-breadcrumb-link_link--initial-underline:focus,
.c5.fi-breadcrumb-link .fi-breadcrumb-link_link--initial-underline:focus-within {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c5.fi-breadcrumb-link .fi-breadcrumb-link_link--initial-underline:hover,
.c5.fi-breadcrumb-link .fi-breadcrumb-link_link--initial-underline:active {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c5.fi-breadcrumb-link .fi-breadcrumb-link_link--current {
  color: hsl(202,7%,40%);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c5.fi-breadcrumb-link .fi-breadcrumb-link_link--current:hover,
.c5.fi-breadcrumb-link .fi-breadcrumb-link_link--current:active {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c5.fi-breadcrumb-link .fi-breadcrumb-link_icon {
  -webkit-transform: translateY(0.2em);
  -ms-transform: translateY(0.2em);
  transform: translateY(0.2em);
  margin: 0 4px;
}

.c5.fi-breadcrumb-link .fi-breadcrumb-link_icon .fi-icon-base-fill {
  fill: hsl(202,7%,40%);
}

<nav
  aria-label="breadcrumb"
  class="c0 fi-breadcrumb c1"
  data-testid="Breadcrumb"
>
  <ol
    class="c2 fi-breadcrumb_list"
  >
    <li
      class="c3 fi-breadcrumb_item"
    >
      <span
        class="c4 c5 fi-breadcrumb-link"
      >
        <a
          class="c6 fi-link c7 fi-breadcrumb-link_link"
          href="/"
        >
          Frontpage
        </a>
        <svg
          aria-hidden="true"
          class="fi-icon c8 fi-breadcrumb-link_icon"
          focusable="false"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            class="fi-icon-base-fill"
            d="M6 18v-2.12l5.902-2.328 3.748-1.492v-.12l-3.748-1.492L6 8.119V6l12 5.015v1.97z"
            fill="#222"
            fill-rule="evenodd"
          />
        </svg>
      </span>
    </li>
    <li
      class="c3 fi-breadcrumb_item"
    >
      <span
        class="c4 c5 fi-breadcrumb-link"
      >
        <a
          class="c6 fi-link c7 fi-breadcrumb-link_link"
          href="/sub"
        >
          Sub page
        </a>
        <svg
          aria-hidden="true"
          class="fi-icon c8 fi-breadcrumb-link_icon"
          focusable="false"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            class="fi-icon-base-fill"
            d="M6 18v-2.12l5.902-2.328 3.748-1.492v-.12l-3.748-1.492L6 8.119V6l12 5.015v1.97z"
            fill="#222"
            fill-rule="evenodd"
          />
        </svg>
      </span>
    </li>
    <li
      class="c3 fi-breadcrumb_item"
    >
      <span
        class="c4 c5 fi-breadcrumb-link"
      >
        <span
          aria-current="location"
          class="c4 fi-breadcrumb-link_link fi-breadcrumb-link_link--current"
        >
          Sub sub page
        </span>
      </span>
    </li>
  </ol>
</nav>
`;
